### ISSUE TITLE
More in-depth "Undocumented*" rules with missing/invalid block tags

### ISSUE THREAD

**Issue by utybo** [detekt/detekt#4395]
## Expected Behavior of the rule

It would be nice if the following rules could go a little more in-depth by also checking for missing or invalid block tags.

### [`comments.UndocumentedPublicClass`](https://detekt.github.io/detekt/comments.html#undocumentedpublicclass)

- Report undocumented (constructor) parameter and type parameter. Each one of them should either have a corresponding `@param` tag in the class' KDoc *or* have its own KDoc. Also report `@param` blocks that do not correspond to any parameter.

```kotlin
/**
 * I am a documentation string
 */
class InvalidBecauseMissingParamDocs<T>(something: String)

/**
 * I am a documentation string
 *
 * @param foobar What the hell is this?
 * @param something My favorite parameter
 * @param T My generic parameter
 */
class ValidBecauseNoParameterNamedFoobar<T>(something: String)

/**
 * I am a documentation string
 *
 * @param something My favorite parameter
 * @param T My generic parameter
 */
class ValidBecauseInClassKDoc<T>(something: String)

// The following example does not apply to generic parameters, as they cannot have attached KDoc directly.
/**
 * I am a documentation string
 */
class ValidBecauseHasItsOwnKDoc(
    /**
     * My favorite parameter
     */
    something: String
)
```

### [`comments.UndocumentedPublicFunction`](https://detekt.github.io/detekt/comments.html#undocumentedpublicfunction)

- Report undocumented parameters and type parameters. Same as for class parameters basically.
  - Potential caveat: Only raise an error if the function itself declares a specific type parameter. Type parameters coming from the class containing the function should be ignored.

```kotlin
/**
 * This is some docs
 */
fun <T> invalidBecauseUndocumentedParameter(whatAmI: String) {
    TODO("Am I a cat?")
}

/**
 * This is some docs
 * 
 * @param whatAmI This is some more docs
 * @param wolf I ran out of ideas
 * @param T My generic type
 */
fun <T> invalidBecauseNoParameterNamedWolf(whatAmI: String) {
    TODO("Am I a cat?")
}

/**
 * This is some docs
 * 
 * @param whatAmI This is some more docs
 * @param T My generic type
 */
fun <T> validBecauseInFunctionKDoc(whatAmI: String) {
    TODO("Am I a cat?")
}

// (Same remark as for class generic type parameters)
/**
 * This is some docs
 */
fun validBecauseHasItsOwnKDoc(
    /**
     * This is some more docs
     */
    whatAmI: String
) {
    TODO("Am I a cat?")
}
```

- Report undocumented return value if the function does not return Unit nor Nothing

```kotlin
/**
 * I don't really do anything special
 */
fun validBecauseReturnsUnit() {
    println("""¯\_(ツ)_/¯""")
}

/**
 * I do a bit of a throw
 *
 * @throws NotImplementedError I'm not done yet
 */
fun validBecauseReturnsNothing(): Nothing = TODO("I throw things around")

/**
 * What do I return?
 */
fun invalidBecauseUndocumentedReturnValue(): Int = 42

/**
 * What do I return?
 * 
 * @return A pretty cool answer
 */
fun validBecauseDocumentedReturnValue(): Int = 42
```

- Report undocumented exceptions if the function reports nothing. I don't think this one will be able to do much other than report aa missing `@throws`/`@exception` tag.

```kotlin
/**
 * If you call this, I will literally throw something.
 */
fun invalidBecauseNothingReturnedAndNoExceptionDocumented(): Nothing = throw Exception("YEET!")

/**
 * If you call this, I will literally throw something.
 * 
 * @exception Exception I'll throw that.
 */
fun validBecauseExceptionTag(): Nothing = throw Exception("YEET!")

/**
 * If you call this, I will literally throw something.
 *
 * @throws Exception I'll throw that.
 */
fun validBecauseThrowsTag(): Nothing = throw Exception("YEET!")
```

- Report missing `@receiver` documentation for extension functions
  - This one may not be as necessary as the other ones, but may be a nice addition.

### [`comments.UndocumentedPublicProperty`](https://detekt.github.io/detekt/comments.html#undocumentedpublicproperty)

* Report missing documentation for property generic type parameters

### Remarks

These options should be optional -- not everyone wants to go *this* in-depth when it comes to documentation.

## Context

This would help tremendously for keeping a high standard of documentation quality for projects that want them -- I don't think any tool actually does any of these checks in Kotlin although they would not be *that* difficult to implement.

The main context is public libraries that wish to maintain high documentation standards. It is my personal belief that the Kotlin ecosystem has a recurring problem of lacking high quality KDoc, mostly because of the tooling not being that great. Adding these more in-depth checks to Detekt would help in that regard in my opinion.

I would also be okay with creating more rules instead of adding these constraints to the existing rules.

Relevant documentation:

* `@param <name>`: https://kotlinlang.org/docs/kotlin-doc.html#param-name
* `@return`: https://kotlinlang.org/docs/kotlin-doc.html#return
* `@receiver`: https://kotlinlang.org/docs/kotlin-doc.html#receiver
* `@throws <class>` and `@exception <class>`: https://kotlinlang.org/docs/kotlin-doc.html#throws-class-exception-class

**Comment by BraisGabin** [comment]
Wow! what a detailed issue! Thanks! I agree, those rules can be improved. And I think that all your suggestions can be added to the curren rules, no need for create new rules.

> * Report undocumented exceptions if the function reports nothing. I don't think this one will be able to do much other than report aa missing `@throws`/`@exception` tag.

About this one I imagine that you want to check that if `Nothing` is the return type at least one `@throws` or `@exception` is declared, right?

--- 

Do you want/can to give it a try and improve those rules yourself?

**Comment by utybo** [comment]
> About this one I imagine that you want to check that if Nothing is the return type at least one @throws or @exception is declared, right?

Yep, that's what I had in mind but I missed it in the report, my apologies.

> Do you want/can to give it a try and improve those rules yourself?

Sure, I can give it a try, but no promises on any actual time window 😅

**Comment by utybo** [comment]
While implementing the changes to the rules, I noticed that the check for functions that have a return type of `Unit` or `Nothing` (especially implicitly) may require a bindingcontext. As I do not want to add type resolution dependent code to a rule that did not have any in the first place, I will probably create a separate rule for this, unless there is a way of determining the return type of a function without a binding context (The Kotlin PSI's (lack of) documentation is horrible for a beginner like me to be honest...)

**Comment by utybo** [comment]
As I've mentioned in the PR, a lot of this is actually already present in the rule `OutdatedDocumentation`. The only thing that needs to be changed is to make a restriction optional, which I'll do in a new PR.

**Comment by BraisGabin** [comment]
Does it have sense to keep `OutdatedDocumentation`? I mean, this confusion was a "valid one". Maybe we shouldn't have a rule called `OutdatedDocumentation` and we should move what it does to other rules. I'm not sure about this, I would like to know what do you think it's the easier way to think about it as an user.

**Comment by utybo** [comment]
As a user, it would probably make more sense to rename it, I agree. "OudatedDocumentation" is a very generic name, and when I first looked at said name, I thought it was related to some versioning documentation (some `@since` or `@sinceKotlinVersion` tag of some sort). The name is made especially weirder by my PR since it actually removes what made it *specifically* about outdatedness.

I think that keeping it as a separate rule makes sense, it just needs a better name imo. Maybe something like:

- `UndocumentedParameter`, consistent with other rules of this kind, but does not actually describe what the rule does as it also checks for consistency
- `MatchingParameterDocumentation`, would be more accurate but maybe somewhat less clear
